File: src/RAG_v2/packages/shared/src/types/mobile.ts
import type { RetrievedDocument } from './chat';

export interface Bookmark {
  id: string;
  session_id: string;
  turn_id: number;
  question: string;
  answer_preview: string;
  answer_snapshot: string;
  sources_snapshot: RetrievedDocument[];
  folder: string;
  note?: string | null;
  created_at: string;
  updated_at?: string;
}

export interface BookmarkFolder {
  name: string;
  count: number;
}

export interface BookmarkCreateRequest {
  session_id: string;
  turn_id: number;
  folder?: string;
  note?: string;
}

export interface BookmarkUpdateRequest {
  folder?: string;
  note?: string;
}

export interface BookmarkFolderRenameRequest {
  new_name: string;
}

export interface FeedbackCreateRequest {
  session_id: string;
  turn_id: number;
  rating: 'up' | 'down';
  category?: 'wrong' | 'incomplete' | 'outdated';
  comment?: string;
}

export interface FeedbackResponse {
  id: string;
  session_id: string;
  turn_id: number;
  rating: 'up' | 'down' | null;
  category?: 'wrong' | 'incomplete' | 'outdated' | null;
  comment?: string | null;
  question: string;
  answer_snapshot: string;
  created_at: string;
  updated_at?: string;
}

export interface FeedbackStats {
  total: number;
  up: number;
  down: number;
  by_category: Record<string, number>;
  recent_days: number;
  with_comment?: number;
}

export interface LookupDocument {
  title: string;
  summary: string;
  source?: string | null;
  date?: string | null;
  url?: string | null;
  collection?: string | null;
  score: number;
  metadata: Record<string, unknown>;
}

export interface SuggestedQuestion {
  question: string;
  category: string;
  popularity: number;
}

export interface NotificationItem {
  id: string;
  title: string;
  body: string;
  type: string;
  related_doc_id?: string | null;
  read: boolean;
  created_at: string;
}

export interface NotificationSubscribeRequest {
  topics: string[];
  expo_push_token: string;
}

export interface NotificationUnsubscribeRequest {
  expo_push_token: string;
  topics?: string[];
}

export interface NotificationUnreadCount {
  unread_count: number;
}

File: src/RAG_v2/schemas/mobile.py
"""Pydantic schemas for mobile-specific feature APIs."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


class BookmarkCreate(BaseModel):
    session_id: str
    turn_id: int = Field(..., ge=1)
    folder: str = Field(default="Chung", min_length=1, max_length=80)
    note: Optional[str] = Field(default=None, max_length=1000)


class BookmarkUpdate(BaseModel):
    folder: Optional[str] = Field(default=None, min_length=1, max_length=80)
    note: Optional[str] = Field(default=None, max_length=1000)


class BookmarkFolderCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=80)


class BookmarkFolderRename(BaseModel):
    new_name: str = Field(..., min_length=1, max_length=80)


class FeedbackCreate(BaseModel):
    session_id: str
    turn_id: int = Field(..., ge=1)
    rating: Literal["up", "down"]
    category: Optional[Literal["wrong", "incomplete", "outdated"]] = None
    comment: Optional[str] = Field(default=None, max_length=1000)


class NotificationSubscribe(BaseModel):
    topics: list[str] = Field(default_factory=list)
    expo_push_token: str = Field(..., min_length=1)


class NotificationUnsubscribe(BaseModel):
    expo_push_token: str = Field(..., min_length=1)
    topics: list[str] = Field(default_factory=list)


class LookupDocument(BaseModel):
    title: str
    summary: str
    collection: str | None = None
    score: float = 0.0
    metadata: dict[str, Any] = Field(default_factory=dict)


class NotificationCreateInternal(BaseModel):
    user_id: str
    title: str
    body: str
    type: str = "update"
    related_doc_id: Optional[str] = None
    read: bool = False
    created_at: datetime
